File: Minimal/src/Brontide.Minimal.Host/ProviderPublisherTrustPolicyDistribution.fs
namespace Brontide.Minimal.Host

open System
open System.Buffers.Binary
open System.IO
open System.Security.Cryptography
open System.Text
open System.Threading
open System.Threading.Tasks

[<StructuralEquality; StructuralComparison>]
type ProviderPublisherTrustPolicyDistributionEndpointId =
    private ProviderPublisherTrustPolicyDistributionEndpointId of string

[<RequireQualifiedAccess>]
module ProviderPublisherTrustPolicyDistributionEndpointId =
    let create value =
        let valid =
            not (String.IsNullOrEmpty value) && value.Length = 64
            && value |> Seq.forall (fun character ->
                (character >= '0' && character <= '9') || (character >= 'A' && character <= 'F'))
        if not valid then invalidArg (nameof value) "A distribution endpoint identity must be an uppercase SHA-256 digest."
        ProviderPublisherTrustPolicyDistributionEndpointId value
    let value (ProviderPublisherTrustPolicyDistributionEndpointId value) = value

type ProviderPublisherTrustPolicyDistributionRequest =
    { Challenge: string
      CurrentSequence: int64
      CurrentPolicyIdentity: ProviderPublisherTrustPolicyId option }

type ProviderPublisherTrustPolicyDistributionResponse =
    { Challenge: string
      CurrentSequence: int64
      CurrentPolicyIdentity: ProviderPublisherTrustPolicyId option
      IssuedAtUnixSeconds: int64
      ExpiresAtUnixSeconds: int64
      Update: ProviderPublisherTrustPolicyUpdate option
      Algorithm: string
      EndpointPublicKeySpkiBase64: string
      SignatureBase64: string }

type IProviderPublisherTrustPolicyDistributionSource =
    abstract FetchAsync:
        ProviderPublisherTrustPolicyDistributionRequest * CancellationToken ->
            Task<ProviderPublisherTrustPolicyDistributionResponse>

type ProviderPublisherTrustPolicyDistributionResult =
    { Code: string
      Current: VerifiedProviderPublisherTrustPolicySnapshot option
      Floor: ProviderPublisherTrustPolicyRecoveryFloor }
    member this.IsApplied = this.Code = "policy-distribution-applied"

[<RequireQualifiedAccess>]
module ProviderPublisherTrustPolicyDistributionManifest =
    let private appendInt32 (output: Stream) value =
        let bytes = Array.zeroCreate<byte> 4
        BinaryPrimitives.WriteInt32BigEndian(bytes.AsSpan(), value)
        output.Write bytes
    let private appendInt64 (output: Stream) value =
        let bytes = Array.zeroCreate<byte> 8
        BinaryPrimitives.WriteInt64BigEndian(bytes.AsSpan(), value)
        output.Write bytes
    let private appendString output (value: string) =
        let bytes = Encoding.UTF8.GetBytes value
        appendInt32 output bytes.Length
        output.Write bytes

    let updateDigest (update: ProviderPublisherTrustPolicyUpdate) =
        if isNull (box update) then nullArg (nameof update)
        use output = new MemoryStream()
        appendString output "CBI39-UPDATE"
        appendInt64 output update.Sequence
        appendInt32 output (if update.PreviousPolicyIdentity.IsSome then 1 else 0)
        update.PreviousPolicyIdentity |> Option.iter (ProviderPublisherTrustPolicyId.value >> appendString output)
        appendString output (ProviderPublisherTrustPolicyId.value update.Policy.Identity)
        let entries = update.Policy.Entries |> List.sortBy (fun entry -> ProviderPublisherKeyId.value entry.PublisherKeyId)
        appendInt32 output entries.Length
        for entry in entries do
            appendString output (ProviderPublisherKeyId.value entry.PublisherKeyId)
            appendString output (if entry.Disposition = Admitted then "admitted" else "revoked")
        appendString output update.Algorithm
        appendString output update.AuthorityPublicKeySpkiBase64
        appendString output update.SignatureBase64
        output.ToArray() |> SHA256.HashData |> Convert.ToHexString

    let encode challenge currentSequence currentPolicyIdentity issuedAt expiresAt update =
        use output = new MemoryStream()
        appendString output "CBI39"
        appendString output challenge
        appendInt64 output currentSequence
        appendInt32 output (if Option.isSome currentPolicyIdentity then 1 else 0)
        currentPolicyIdentity |> Option.iter (ProviderPublisherTrustPolicyId.value >> appendString output)
        appendInt64 output issuedAt
        appendInt64 output expiresAt
        appendInt32 output (if Option.isSome update then 1 else 0)
        update |> Option.iter (updateDigest >> appendString output)
        output.ToArray()

type ProviderPublisherTrustPolicyDistributionClient(
    registry: DurableProviderPublisherTrustPolicyRegistry,
    endpointIdentity: ProviderPublisherTrustPolicyDistributionEndpointId) =

    do
        if isNull (box registry) then nullArg (nameof registry)
        if isNull (box endpointIdentity) then invalidArg (nameof endpointIdentity) "An endpoint identity is required."

    let maximumCharacters = 1024L * 1024L
    let maximumValidity = TimeSpan.FromMinutes 15.0
    let maximumFutureSkew = TimeSpan.FromSeconds 30.0

    let result code = { Code = code; Current = registry.Current; Floor = registry.Floor }

    let cursorMatches (request: ProviderPublisherTrustPolicyDistributionRequest) =
        match registry.Current with
        | None -> request.CurrentSequence = 0L && request.CurrentPolicyIdentity.IsNone
        | Some current ->
            request.CurrentSequence = current.Sequence
            && request.CurrentPolicyIdentity = Some current.Policy.Identity

    let bounded (response: ProviderPublisherTrustPolicyDistributionResponse) =
        if isNull (box response) || response.Algorithm <> "ECDSA-P256-SHA256" || response.Challenge.Length <> 64
            || response.Challenge |> Seq.exists (Uri.IsHexDigit >> not) then false
        else
            match response.Update with
            | Some update when isNull (box update.Policy) || update.Policy.Entries.Length > 4096
                -> false
            | update ->
                let mutable characters =
                    int64 response.Challenge.Length + int64 response.Algorithm.Length
                    + int64 response.EndpointPublicKeySpkiBase64.Length + int64 response.SignatureBase64.Length
                match update with
                | Some value ->
                    characters <- characters + int64 (
                        value.Algorithm.Length + value.AuthorityPublicKeySpkiBase64.Length + value.SignatureBase64.Length)
                    characters <- characters + (value.Policy.Entries |> List.sumBy (fun entry ->
                        int64 (ProviderPublisherKeyId.value entry.PublisherKeyId).Length + 8L))
                | None -> ()
                characters <= maximumCharacters

    member _.SynchronizeAsync(
        source: IProviderPublisherTrustPolicyDistributionSource,
        now: DateTimeOffset,
        timeout: TimeSpan,
        cancellationToken: CancellationToken) = task {
        if isNull (box source) then nullArg (nameof source)
        if timeout <= TimeSpan.Zero || timeout > TimeSpan.FromMinutes 1.0 then
            invalidArg (nameof timeout) "The timeout must be positive and no greater than one minute."
        let current = registry.Current
        let request =
            { Challenge = RandomNumberGenerator.GetBytes 32 |> Convert.ToHexString
              CurrentSequence = current |> Option.map _.Sequence |> Option.defaultValue 0L
              CurrentPolicyIdentity = current |> Option.map _.Policy.Identity }
        use timeoutSource = CancellationTokenSource.CreateLinkedTokenSource cancellationToken
        timeoutSource.CancelAfter timeout
        let! fetched = task {
            try
                let! response = source.FetchAsync(request, timeoutSource.Token).WaitAsync(timeout, cancellationToken)
                return Ok response
            with
            | :? TimeoutException -> return Error "policy-distribution-timeout"
            | :? OperationCanceledException when cancellationToken.IsCancellationRequested ->
                return Error "policy-distribution-canceled"
            | :? OperationCanceledException -> return Error "policy-distribution-timeout"
            | error when not (error :? OutOfMemoryException) ->
                return Error "policy-distribution-transport-failed"
        }
        match fetched with
        | Error code -> return result code
        | Ok response when not (bounded response) -> return result "policy-distribution-response-invalid"
        | Ok response ->
            let verification =
                try
                    let publicKey = Convert.FromBase64String response.EndpointPublicKeySpkiBase64
                    let signature = Convert.FromBase64String response.SignatureBase64
                    let suppliedEndpoint =
                        publicKey |> SHA256.HashData |> Convert.ToHexString
                        |> ProviderPublisherTrustPolicyDistributionEndpointId.create
                    if suppliedEndpoint <> endpointIdentity then Error "policy-distribution-endpoint-mismatch"
                    else
                        use verifier = ECDsa.Create()
                        let mutable bytesRead = 0
                        verifier.ImportSubjectPublicKeyInfo(publicKey.AsSpan(), &bytesRead)
                        let curve = verifier.ExportParameters(false).Curve.Oid.Value |> Option.ofObj
                        if bytesRead <> publicKey.Length || verifier.KeySize <> 256
                            || curve <> Some "1.2.840.10045.3.1.7" then
                            Error "policy-distribution-response-invalid"
                        elif verifier.VerifyData(
                            ProviderPublisherTrustPolicyDistributionManifest.encode
                                response.Challenge response.CurrentSequence response.CurrentPolicyIdentity
                                response.IssuedAtUnixSeconds response.ExpiresAtUnixSeconds response.Update,
                            signature, HashAlgorithmName.SHA256, DSASignatureFormat.Rfc3279DerSequence) then Ok ()
                        else Error "policy-distribution-signature-invalid"
                with
                | :? FormatException
                | :? CryptographicException
                | :? ArgumentException
                | :? NullReferenceException -> Error "policy-distribution-response-invalid"
            match verification with
            | Error code -> return result code
            | Ok () when response.Challenge <> request.Challenge ->
                return result "policy-distribution-challenge-mismatch"
            | Ok () when response.CurrentSequence <> request.CurrentSequence
                || response.CurrentPolicyIdentity <> request.CurrentPolicyIdentity ->
                return result "policy-distribution-cursor-mismatch"
            | Ok () ->
                let freshness =
                    try
                        let issued = DateTimeOffset.FromUnixTimeSeconds response.IssuedAtUnixSeconds
                        let expires = DateTimeOffset.FromUnixTimeSeconds response.ExpiresAtUnixSeconds
                        (issued <= now || issued - now <= maximumFutureSkew) && expires > now && expires > issued
                        && expires - issued <= maximumValidity
                    with :? ArgumentOutOfRangeException -> false
                if not freshness then return result "policy-distribution-stale"
                elif not (cursorMatches request) then return result "policy-distribution-superseded"
                else
                    match response.Update with
                    | None -> return result "policy-distribution-current"
                    | Some update ->
                        let applied = registry.Apply update
                        return { Code = if applied.IsApplied then "policy-distribution-applied" else applied.Code
                                 Current = applied.Current; Floor = applied.Floor }
    }

File: Reference/src/Brontide.Reference.Studio/ProviderPublisherTrustPolicyDistribution.cs
using System.Buffers.Binary;
using System.Security.Cryptography;
using System.Text;

namespace Brontide.Reference.Studio;

public readonly record struct ProviderPublisherTrustPolicyDistributionEndpointId
{
    private ProviderPublisherTrustPolicyDistributionEndpointId(string value) => Value = value;
    public string Value { get; }

    public static ProviderPublisherTrustPolicyDistributionEndpointId Create(string value)
    {
        ArgumentNullException.ThrowIfNull(value);
        if (!ContentAddressedProviderStore.IsDigest(value))
            throw new ArgumentException("A distribution endpoint identity must be an uppercase SHA-256 digest.", nameof(value));
        return new(value);
    }
}

public sealed record ProviderPublisherTrustPolicyDistributionRequest(
    string Challenge,
    long CurrentSequence,
    ProviderPublisherTrustPolicyId? CurrentPolicyIdentity);

public sealed record ProviderPublisherTrustPolicyDistributionResponse(
    string Challenge,
    long CurrentSequence,
    ProviderPublisherTrustPolicyId? CurrentPolicyIdentity,
    long IssuedAtUnixSeconds,
    long ExpiresAtUnixSeconds,
    ProviderPublisherTrustPolicyUpdate? Update,
    string Algorithm,
    string EndpointPublicKeySpkiBase64,
    string SignatureBase64);

public interface IProviderPublisherTrustPolicyDistributionSource
{
    Task<ProviderPublisherTrustPolicyDistributionResponse> FetchAsync(
        ProviderPublisherTrustPolicyDistributionRequest request,
        CancellationToken cancellationToken);
}

public sealed record ProviderPublisherTrustPolicyDistributionResult(
    string Code,
    VerifiedProviderPublisherTrustPolicySnapshot? Current,
    ProviderPublisherTrustPolicyRecoveryFloor Floor)
{
    public bool IsApplied => Code == "policy-distribution-applied";
}

public static class ProviderPublisherTrustPolicyDistributionManifest
{
    public static byte[] Encode(
        string challenge,
        long currentSequence,
        ProviderPublisherTrustPolicyId? currentPolicyIdentity,
        long issuedAtUnixSeconds,
        long expiresAtUnixSeconds,
        ProviderPublisherTrustPolicyUpdate? update)
    {
        using var output = new MemoryStream();
        Append(output, "CBI39");
        Append(output, challenge);
        Append(output, currentSequence);
        Append(output, currentPolicyIdentity.HasValue ? 1 : 0);
        if (currentPolicyIdentity.HasValue) Append(output, currentPolicyIdentity.Value.Value);
        Append(output, issuedAtUnixSeconds);
        Append(output, expiresAtUnixSeconds);
        Append(output, update is null ? 0 : 1);
        if (update is not null) Append(output, UpdateDigest(update));
        return output.ToArray();
    }

    public static string UpdateDigest(ProviderPublisherTrustPolicyUpdate update)
    {
        ArgumentNullException.ThrowIfNull(update);
        using var output = new MemoryStream();
        Append(output, "CBI39-UPDATE");
        Append(output, update.Sequence);
        Append(output, update.PreviousPolicyIdentity.HasValue ? 1 : 0);
        if (update.PreviousPolicyIdentity.HasValue) Append(output, update.PreviousPolicyIdentity.Value.Value);
        Append(output, update.Policy.Identity.Value);
        var entries = update.Policy.Entries.OrderBy(entry => entry.PublisherKeyId.Value, StringComparer.Ordinal).ToArray();
        Append(output, entries.Length);
        foreach (var entry in entries)
        {
            Append(output, entry.PublisherKeyId.Value);
            Append(output, entry.Disposition == ProviderPublisherTrustDisposition.Admitted ? "admitted" : "revoked");
        }
        Append(output, update.Algorithm);
        Append(output, update.AuthorityPublicKeySpkiBase64);
        Append(output, update.SignatureBase64);
        return Convert.ToHexString(SHA256.HashData(output.ToArray()));
    }

    private static void Append(Stream output, int value)
    {
        Span<byte> bytes = stackalloc byte[4];
        BinaryPrimitives.WriteInt32BigEndian(bytes, value);
        output.Write(bytes);
    }

    private static void Append(Stream output, long value)
    {
        Span<byte> bytes = stackalloc byte[8];
        BinaryPrimitives.WriteInt64BigEndian(bytes, value);
        output.Write(bytes);
    }

    private static void Append(Stream output, string value)
    {
        var bytes = Encoding.UTF8.GetBytes(value);
        Append(output, bytes.Length);
        output.Write(bytes);
    }
}

public sealed class ProviderPublisherTrustPolicyDistributionClient
{
    private const int MaxEncodedCharacters = 1024 * 1024;
    private static readonly TimeSpan MaximumValidity = TimeSpan.FromMinutes(15);
    private static readonly TimeSpan MaximumFutureSkew = TimeSpan.FromSeconds(30);
    private readonly DurableProviderPublisherTrustPolicyRegistry registry;
    private readonly ProviderPublisherTrustPolicyDistributionEndpointId endpointIdentity;

    public ProviderPublisherTrustPolicyDistributionClient(
        DurableProviderPublisherTrustPolicyRegistry registry,
        ProviderPublisherTrustPolicyDistributionEndpointId endpointIdentity)
    {
        ArgumentNullException.ThrowIfNull(registry);
        if (endpointIdentity.Value is null) throw new ArgumentException("An endpoint identity is required.", nameof(endpointIdentity));
        this.registry = registry;
        this.endpointIdentity = endpointIdentity;
    }

    public async Task<ProviderPublisherTrustPolicyDistributionResult> SynchronizeAsync(
        IProviderPublisherTrustPolicyDistributionSource source,
        DateTimeOffset now,
        TimeSpan timeout,
        CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(source);
        if (timeout <= TimeSpan.Zero || timeout > TimeSpan.FromMinutes(1))
            throw new ArgumentOutOfRangeException(nameof(timeout));

        var current = registry.Current;
        var request = new ProviderPublisherTrustPolicyDistributionRequest(
            Convert.ToHexString(RandomNumberGenerator.GetBytes(32)),
            current?.Sequence ?? 0,
            current?.Policy.Identity);
        ProviderPublisherTrustPolicyDistributionResponse response;
        using var timeoutSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
        timeoutSource.CancelAfter(timeout);
        try
        {
            response = await source.FetchAsync(request, timeoutSource.Token)
                .WaitAsync(timeout, cancellationToken).ConfigureAwait(false);
        }
        catch (TimeoutException)
        {
            return Result("policy-distribution-timeout");
        }
        catch (OperationCanceledException) when (cancellationToken.IsCancellationRequested)
        {
            return Result("policy-distribution-canceled");
        }
        catch (OperationCanceledException)
        {
            return Result("policy-distribution-timeout");
        }
        catch (Exception exception) when (exception is not OutOfMemoryException)
        {
            return Result("policy-distribution-transport-failed");
        }

        if (!IsBounded(response)) return Result("policy-distribution-response-invalid");
        byte[] publicKey;
        byte[] signature;
        try
        {
            publicKey = Convert.FromBase64String(response.EndpointPublicKeySpkiBase64);
            signature = Convert.FromBase64String(response.SignatureBase64);
            var suppliedIdentity = ProviderPublisherTrustPolicyDistributionEndpointId.Create(
                Convert.ToHexString(SHA256.HashData(publicKey)));
            if (suppliedIdentity != endpointIdentity) return Result("policy-distribution-endpoint-mismatch");
            using var verifier = ECDsa.Create();
            verifier.ImportSubjectPublicKeyInfo(publicKey, out var bytesRead);
            var parameters = verifier.ExportParameters(false);
            if (bytesRead != publicKey.Length || verifier.KeySize != 256
                || parameters.Curve.Oid.Value != "1.2.840.10045.3.1.7")
                return Result("policy-distribution-response-invalid");
            if (!verifier.VerifyData(ProviderPublisherTrustPolicyDistributionManifest.Encode(
                    response.Challenge, response.CurrentSequence, response.CurrentPolicyIdentity,
                    response.IssuedAtUnixSeconds, response.ExpiresAtUnixSeconds, response.Update),
                signature, HashAlgorithmName.SHA256, DSASignatureFormat.Rfc3279DerSequence))
                return Result("policy-distribution-signature-invalid");
        }
        catch (Exception exception) when (exception is FormatException or CryptographicException
            or ArgumentException or NullReferenceException)
        {
            return Result("policy-distribution-response-invalid");
        }

        if (response.Challenge != request.Challenge) return Result("policy-distribution-challenge-mismatch");
        if (response.CurrentSequence != request.CurrentSequence
            || response.CurrentPolicyIdentity != request.CurrentPolicyIdentity)
            return Result("policy-distribution-cursor-mismatch");
        try
        {
            var issued = DateTimeOffset.FromUnixTimeSeconds(response.IssuedAtUnixSeconds);
            var expires = DateTimeOffset.FromUnixTimeSeconds(response.ExpiresAtUnixSeconds);
            if ((issued > now && issued - now > MaximumFutureSkew)
                || expires <= now || expires <= issued || expires - issued > MaximumValidity)
                return Result("policy-distribution-stale");
        }
        catch (ArgumentOutOfRangeException)
        {
            return Result("policy-distribution-stale");
        }
        if (!CursorMatches(request)) return Result("policy-distribution-superseded");
        if (response.Update is null) return Result("policy-distribution-current");

        var applied = registry.Apply(response.Update);
        return new(applied.IsApplied ? "policy-distribution-applied" : applied.Code, applied.Current, applied.Floor);
    }

    private bool CursorMatches(ProviderPublisherTrustPolicyDistributionRequest request)
    {
        var current = registry.Current;
        return (current?.Sequence ?? 0) == request.CurrentSequence
            && current?.Policy.Identity == request.CurrentPolicyIdentity;
    }

    private ProviderPublisherTrustPolicyDistributionResult Result(string code) =>
        new(code, registry.Current, registry.Floor);

    private static bool IsBounded(ProviderPublisherTrustPolicyDistributionResponse? response)
    {
        if (response is null || response.Challenge is null || response.Algorithm != "ECDSA-P256-SHA256"
            || response.EndpointPublicKeySpkiBase64 is null || response.SignatureBase64 is null
            || response.Challenge.Length != 64 || response.Challenge.Any(character => !Uri.IsHexDigit(character)))
            return false;
        var update = response.Update;
        if (update is not null && (update.Policy is null || update.Policy.Entries.Count > 4096
            || update.Algorithm is null || update.AuthorityPublicKeySpkiBase64 is null || update.SignatureBase64 is null))
            return false;
        long characters = (long)response.Challenge.Length + response.Algorithm.Length
            + response.EndpointPublicKeySpkiBase64.Length + response.SignatureBase64.Length;
        if (update is not null)
        {
            characters += update.Algorithm.Length + update.AuthorityPublicKeySpkiBase64.Length + update.SignatureBase64.Length;
            characters += update.Policy.Entries.Sum(entry => (long)entry.PublisherKeyId.Value.Length + 8);
        }
        return characters <= MaxEncodedCharacters;
    }
}
